### PR TITLE
Bugreport/concat exception

### PR DESCRIPTION
**Pull request by u3r**
I encountered a bug when concatenating RX-ified Mongo queries.
This pullrequest demonstrates the problem: 
One unittest will crash with an exception but continue running
The other (with an `.doOnException(..log..))` will behave as expected even though the `onException` is never hit.

### CODE REVIEW

**Comment by karianna** [comment]
@u3r Can you rebase (there are conflicts)?

**Comment by u3r** [comment]
Hi @karianna,
I did not rebase as
a) I cannot access the current snapshot repositories so I can't verify it compiling
b) It is meant as support for #153 - please either drop the modification of pom.xml or rebase onto the 3.5.1 tag to check the problem.

**Comment by karianna** [comment]
Are you not able to test against the 3.5.2 stack?

**Comment by u3r** [comment]
No, I've not... might have been only the company proxy/artifact storage not being able to access the latest artifacts.
Just try out the test without the .pom if 3.5.2 builds on your machine... should be a matter of minutes.
The test is not fit for continuous integration anyway since the thread hangs until vertx-timout (120minutes).

File: src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  private JsonObject createTestDocument( String deviceId, OffsetDateTime timeStamp ) {
+    JsonObject document = new JsonObject()
+      .put( "deviceId", deviceId )
+      .put( "timestamp", timeStamp.toInstant().toEpochMilli() );
+    return document;
+  }
+
+  public Single<MongoClientBulkWriteResult> insertDocuments( final String collection, List<JsonObject> documents ) {
```
Comment (karianna) [inline src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java]: Could be private

File: src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import io.vertx.rxjava.core.Vertx;
+import rx.Observable;
+import rx.Single;
+import rx.exceptions.Exceptions;
+
+public class MongoConcatWithTest extends MongoTestBase {
+
+  private MongoClient mongoClient;
+
+  private int messageCount = 55;
```
Comment (karianna) [inline src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java]: could be consts e.g. add static and capitalise variable names

File: src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  private int bulkWriteSize = 10;
+
+  private OffsetDateTime now = OffsetDateTime.of( 2018, 1, 4, 10, 25, 0, 0, ZoneOffset.UTC );
```
Comment (karianna) [inline src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java]: If this OffsetDateTime is truly a constant then make it one, perhaps change variable name as well as it's not really 'now'

File: src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      .concatWith( selectItemsOccurringAfter( "MockCollectionName1", "123", now.minusMinutes( 80 ).toInstant() )
+        // IMPORTANT: DO NOT EVER REMOVE THIS or you will get `java.lang.IllegalStateException: more produced than requested`
+        .doOnError( t -> { // when using this with concatWith()
+          System.out.println( "Error while selecting items: " + t.getMessage() );
+          throw Exceptions.propagate( t );
+        } )
+      )
+      .buffer( bulkWriteSize )
+      .concatMap( list -> Observable.just( list.size() ) ) // normally this is a bulk write to mongo
+      .reduce( ( a, b ) -> a + b )
```
Comment (karianna) [inline src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java]: maybe use more descriptive variables than a and b

File: src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      .first();
+
+    // THEN
+    assertEquals( messageCount, updateCount.intValue() );
+  }
+
+  private Observable<Integer> fakeBulkUpdateReturningUpdateCount( Observable<List<JsonObject>> input ) {
+    return input
+      .concatMap( list -> Observable.just( list.size() ) ) // normally this is a bulk write to mongo
+      .reduce( ( a, b ) -> a + b );
```
Comment (karianna) [inline src/test/java/io/vertx/rxjava/ext/mongo/MongoConcatWithTest.java]: more meaningful variable names

**Comment by karianna** [comment]
OK so these tests are useful, thank you!  I guess we need to solve the underlying fault though else we'll get that hanging test.

**Comment by tsegismont** [comment]
@karianna looking into this

**Comment by tsegismont** [comment]
This seems to be an issue with the RxJava 1 adapter for `ReadStream`/`Observable`. The same code with RxJava 2 and `FLowable` works fine.

**Comment by vietj** [comment]
I plan soon to PR with the new read stream adapter, this bug might be fixed with this new feature

**Comment by tsegismont** [comment]
@vietj will try the test with the changes and keep you informed

**Comment by tsegismont** [comment]
@vietj I can't find your branch in Vert.x core, have you shared it?

**Comment by vietj** [comment]
no I haven't, I will soon

**Comment by vietj** [comment]
there are three branches, one for core, one for rx and one for mongo (at least for testing)

**Comment by tsegismont** [comment]
Closing as #153 is fixed in master

**Pull request closed without merging**